src/lib.rs

```diff
@@ -1,70 +1,103 @@
-use image::{imageops::overlay, load_from_memory, ImageError, RgbaImage};
-use rand::random;
+use image::{imageops::overlay, ImageError, DynamicImage};
 use std::path::Path;
+use rand::prelude::*;
+use std::time::Instant;
 
 mod emoji;
 
-fn sums_chunked(samples_a: &[u8], samples_b: &[u8]) -> (i32, i32, i32) {
+
+fn measure_dist_chunks(samples_a: &[u8], samples_b: &[u8]) -> i64 {
     samples_a
-        .chunks_exact(3)
-        .zip(samples_b.chunks_exact(3))
-        .fold((0, 0, 0), |(r, g, b), (p_a, p_b)| {
-            (
-                r + (p_a[0] as i32 - p_b[0] as i32).abs(),
-                g + (p_a[1] as i32 - p_b[1] as i32).abs(),
-                b + (p_a[2] as i32 - p_b[2] as i32).abs(),
-            )
+        .chunks_exact(1)
+        .zip(samples_b.chunks_exact(1))
+        .fold(0, |rgba, (p_a, p_b)| {
+                rgba + (p_a[0] as i64 - p_b[0] as i64).abs()
         })
 }
 
+fn subimage_compare(image_a: &DynamicImage, image_b: &DynamicImage, x: u32, y: u32 ) -> i64 {
+    let sub_image_a = image_a.crop_imm(x,y,16,16).to_rgb();
+    let sub_image_b = image_b.crop_imm(x,y,16,16).to_rgb();
+    measure_dist_chunks(&sub_image_a, &sub_image_b)
+}
+
 pub fn generate_image(
-    image_buffer: &[u8],
+    image_buffer: &DynamicImage,
     iterations: u64,
     save_progress: bool,
     path: &Path,
 ) -> Result<Vec<u8>, ImageError> {
+    let now = Instant::now();
+
     let mut emoji_cache = emoji::EmojiCache::new();
-    let orig = load_from_memory(image_buffer)?.into_rgba();
-    let (width, height) = orig.dimensions();
-    let mut new_img = RgbaImage::new(width, height);
-    let mut dist = sums_chunked(&orig, &new_img);
 
-    for _ in 0..iterations {
+    let image_buffer_rgb = image_buffer.clone().to_rgb();
+    let (width, height) = image_buffer_rgb.dimensions();
+    let canvas_size = width * height;
+    let mut rng = thread_rng();
+    let mut new_img = DynamicImage::new_rgb16(width, height);
+    println!("image is {} by {} pixels", width, height);
+    let mut dist = measure_dist_chunks(&image_buffer_rgb, &new_img.to_rgb());
+    println!("dist is {}", dist);
+
+    let mut placed_count = 0;
+    for _  in 0..canvas_size/20 {
         let e = emoji_cache.get_emoji();
-        let x: u32 = random::<u32>() % width;
-        let y: u32 = random::<u32>() % height;
+        let x: u32 = (1..width).choose(&mut rng).unwrap();
+        let y: u32 = (1..height).choose(&mut rng).unwrap();
```

Comment:
i would maybe declare something like
`let width_range = 1..width` above, rather than doing this every iteration.
also, should it be `0..width`?

Reply:
I have been trying to do the single declare. 
So far I've only been able to do that it I clone anyway.
I've tried to see what is the most performant and so far this seems more performant in this loop than the random::<u32> method  
Working on it